pkg/redis/client.go

```diff
@@ -0,0 +1,188 @@
+package redis
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+)
+
+// ErrNotFound indicates that the record is not found in redis.
+var (
+	ErrNotFound = errors.New("redis: not found")
+)
+
+// Client is the public interface so that implementations can be mocked or swapped.
+type Client interface {
+	GetSandboxBySessionID(ctx context.Context, sessionID string) (*Sandbox, error)
+	SetSessionLockIfAbsent(ctx context.Context, sessionID string, ttl time.Duration) (bool, error)
+	BindSessionWithSandbox(ctx context.Context, sessionID string, sb *Sandbox, ttl time.Duration) error
+	DeleteSessionBySandboxIDTx(ctx context.Context, sandboxID string) error
+}
+
+// client is a go-redis based implementation; it is internal and only exposed via Client.
+type client struct {
+	rdb           *redis.Client
+	sessionPrefix string
+	sandboxPrefix string
+	lockPrefix    string
+}
+
+// Option configures client behavior (e.g. key prefixes).
+type Option func(*client)
+
+// WithKeyPrefixes sets custom key prefixes for session / sandbox / lock.
+func WithKeyPrefixes(session, sandbox, lock string) Option {
+	return func(c *client) {
+		c.sessionPrefix = session
+		c.sandboxPrefix = sandbox
+		c.lockPrefix = lock
+	}
+}
+
+// NewClient creates a new Client using the given go-redis client.
+func NewClient(rdb *redis.Client, opts ...Option) Client {
+	c := &client{
+		rdb:           rdb,
+		sessionPrefix: "session:",
+		sandboxPrefix: "sandbox:",
+		lockPrefix:    "session_lock:",
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	return c
+}
+
+func (c *client) sessionKey(sessionID string) string {
+	return c.sessionPrefix + sessionID
+}
+
+func (c *client) sandboxKey(sandboxID string) string {
+	return c.sandboxPrefix + sandboxID
+}
+
+func (c *client) lockKey(sessionID string) string {
+	return c.lockPrefix + sessionID
+}
+
+// --------- Client implementation ---------
+
+// GetSandboxBySessionID looks up the sandbox bound to the given session ID.
+// Underlying redis command: GET session:{sessionID} -> Sandbox(JSON).
+func (c *client) GetSandboxBySessionID(ctx context.Context, sessionID string) (*Sandbox, error) {
+	key := c.sessionKey(sessionID)
+
+	b, err := c.rdb.Get(ctx, key).Bytes()
+	if err == redis.Nil {
+		return nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("GetSandboxBySessionID: redis GET %s: %w", key, err)
+	}
+
+	var sb Sandbox
+	if err := json.Unmarshal(b, &sb); err != nil {
+		return nil, fmt.Errorf("GetSandboxBySessionID: unmarshal sandbox: %w", err)
+	}
+	return &sb, nil
+}
+
+// SetSessionLockIfAbsent tries to acquire a lock for the given session ID.
+// Underlying redis command: SETNX session_lock:{sessionID} 1 EX ttl.
+func (c *client) SetSessionLockIfAbsent(ctx context.Context, sessionID string, ttl time.Duration) (bool, error) {
```

Comment:
since you explicitly delete the session_lock:{sessionID}, is it still necessary to provide ttl